excise/utils.py

```diff
@@ -0,0 +1,554 @@
+        content = response.content if response else "Unable to find the response"
+        logger.exception(
+            "Unable to decode the response from Avatax Excise. Response: %s", content
+        )
+        return {}
+    return json_response  # type: ignore
+
+
+@dataclass
+class TransactionLine:
+    InvoiceLine: Optional[int]
+    ProductCode: str
+    UnitPrice: Optional[Decimal]
+    UnitOfMeasure: Optional[str]
+    BilledUnits: Optional[Decimal]
+    LineAmount: Optional[Decimal]
+    AlternateUnitPrice: Optional[Decimal]
+    TaxIncluded: bool
+    UnitQuantity: Optional[int]
+    UnitQuantityUnitOfMeasure: Optional[str]
+    DestinationCountryCode: str
+    """ISO 3166-1 alpha-3 code"""
+    DestinationJurisdiction: str
+    DestinationAddress1: Optional[str]
+    DestinationAddress2: Optional[str]
+    DestinationCounty: Optional[str]
+    DestinationCity: str
+    DestinationPostalCode: str
+    SaleCountryCode: str
+    SaleAddress1: Optional[str]
+    SaleAddress2: Optional[str]
+    SaleJurisdiction: str
+    SaleCounty: Optional[str]
+    SaleCity: str
+    SalePostalCode: str
+
+    OriginCountryCode: Optional[str] = None
+    OriginJurisdiction: Optional[str] = None
+    OriginCounty: Optional[str] = None
+    OriginCity: Optional[str] = None
+    OriginPostalCode: Optional[str] = None
+    OriginAddress1: Optional[str] = None
+    OriginAddress2: Optional[str] = None
+
+    UserData: Optional[str] = None
+    CustomString1: Optional[str] = None
+    CustomString2: Optional[str] = None
+    CustomString3: Optional[str] = None
+    CustomNumeric1: Optional[Decimal] = None
+    CustomNumeric2: Optional[Decimal] = None
+    CustomNumeric3: Optional[Decimal] = None
+
+
+@dataclass
+class TransactionCreateRequestData:
+    EffectiveDate: str
+    InvoiceDate: str
+    TitleTransferCode: str
+    TransactionType: str
+    TransactionLines: List[TransactionLine]
+    InvoiceNumber: Optional[str] = None
+
+
+def generate_request_data(
+    transaction_type: str, lines: List[TransactionLine], invoice_number: Optional[str]
+):
+    today_date = str(date.today())  # Does not seem timezone safe
+    data = TransactionCreateRequestData(
+        EffectiveDate=today_date,
+        InvoiceDate=today_date,
+        InvoiceNumber=invoice_number,
+        TitleTransferCode="DEST",
+        TransactionType=transaction_type,
+        TransactionLines=lines,
+    )
+
+    return data
+
+
+def append_line_to_data(
+    data: List[TransactionLine],
+    line_id: int,
+    quantity: int,
+    amount: Decimal,
+    tax_included: bool,
+    variant: "ProductVariant",
+    shipping_address: "Address",
+    variant_channel_listing: "ProductVariantChannelListing",
+):
+    """Abstract line data regardless of Checkout or Order."""
+    stock = variant.stocks.for_country_and_channel(
+        shipping_address.country, variant_channel_listing.channel.slug
+    ).first()
+    if stock:
+        warehouse_address = stock.warehouse.address
```

Comment:
I'm not sure that we should fetch the address for each line separately? It cost us 1 DB request per line.